### PR TITLE
Add support for data poisoning attacks

### PR DESCRIPTION
**Pull request by slivering**
# Problem

Unlike gradient attacks, label flipping is a data poisoning attack, which is why it gets special treatment in the ByzFL implementation. However, this upstream implementation of label flipping is **not modular**. In order to perform label flipping, the attacker needs to have data. ByzFL currently works around the `ByzantineClient` which is designed to perform gradient attacks, not poisoning attacks.

## How the upstream ByzFL handles label flipping

With `LabelFlipping`, the attacker is **omniscient**  and has access to the features and labels of the clients. The attacker first computes the gradients with label flipping applied to honest clients' data, and then sends the mean gradient to the server.
But because a `ByzantineClient` has no knowledge of the actual data, **computing the gradients with label flipping has to be performed by the honest `Client` instances**. This is not a long-term solution for other types of data poisoning.

# Solution

Instead of working around a `ByzantineClient`, we create a `PoisoningClient(Client)` class which provides a generic method to perform data poisoning. This client has the **same status as an honest `Client`**: it is provided a share of the training dataset and poisons a fraction `p` of it.

A poisoning attack can be any callable with the signature `(model, inputs, targets) -> (inputs, targets)`.

This client can now perform any type of label flipping, including _dynamic_ label flipping since it has **access to the model weights**. Furthermore, adding a `PoisoningClient` paves the way for more types of non-omniscient data poisoning attacks, such as gradient inversion.

### CODE REVIEW

**Comment by Ayoub-Gourari** [comment]
Approuvé
